--- drivers/net/wireless/ralink/rt2x00/rt2x00link.c
+++ drivers/net/wireless/ralink/rt2x00/rt2x00link.c
@@ -390,45 +390,28 @@
 					     &link->work, LINK_TUNE_INTERVAL);
 }
 
 void rt2x00link_start_watchdog(struct rt2x00_dev *rt2x00dev)
 {
 	struct link *link = &rt2x00dev->link;
+	
+	return;
 
 	if (test_bit(DEVICE_STATE_PRESENT, &rt2x00dev->flags) &&
 	    rt2x00dev->ops->lib->watchdog)
 		ieee80211_queue_delayed_work(rt2x00dev->hw,
 					     &link->watchdog_work,
 					     WATCHDOG_INTERVAL);
 }
 
 void rt2x00link_stop_watchdog(struct rt2x00_dev *rt2x00dev)
 {
-	cancel_delayed_work_sync(&rt2x00dev->link.watchdog_work);
-}
-
-static void rt2x00link_watchdog(struct work_struct *work)
-{
-	struct rt2x00_dev *rt2x00dev =
-	    container_of(work, struct rt2x00_dev, link.watchdog_work.work);
-	struct link *link = &rt2x00dev->link;
-
-	/*
-	 * When the radio is shutting down we should
-	 * immediately cease the watchdog monitoring.
-	 */
-	if (!test_bit(DEVICE_STATE_ENABLED_RADIO, &rt2x00dev->flags))
-		return;
-
-	rt2x00dev->ops->lib->watchdog(rt2x00dev);
-
-	if (test_bit(DEVICE_STATE_PRESENT, &rt2x00dev->flags))
-		ieee80211_queue_delayed_work(rt2x00dev->hw,
-					     &link->watchdog_work,
-					     WATCHDOG_INTERVAL);
+	return;
+	
+	cancel_delayed_work_sync(&rt2x00dev->link.watchdog_work);
 }
 
 void rt2x00link_register(struct rt2x00_dev *rt2x00dev)
 {
-	INIT_DELAYED_WORK(&rt2x00dev->link.watchdog_work, rt2x00link_watchdog);
+	//INIT_DELAYED_WORK(&rt2x00dev->link.watchdog_work, rt2x00link_watchdog);
 	INIT_DELAYED_WORK(&rt2x00dev->link.work, rt2x00link_tuner);
 }
--- drivers/net/wireless/ralink/rt2x00/rt2x00lib.h
+++ drivers/net/wireless/ralink/rt2x00/rt2x00lib.h
@@ -25,17 +25,17 @@
 #ifndef RT2X00LIB_H
 #define RT2X00LIB_H
 
 /*
  * Interval defines
  */
-#define WATCHDOG_INTERVAL	round_jiffies_relative(HZ)
-#define LINK_TUNE_SECONDS	1
+#define WATCHDOG_INTERVAL	round_jiffies_relative(180 * HZ)
+#define LINK_TUNE_SECONDS	2
 #define LINK_TUNE_INTERVAL	round_jiffies_relative(LINK_TUNE_SECONDS * HZ)
 #define AGC_SECONDS		4
-#define VCO_SECONDS		10
+#define VCO_SECONDS		180
 
 /*
  * rt2x00_rate: Per rate device information
  */
 struct rt2x00_rate {
 	unsigned short flags;
--- drivers/net/wireless/ralink/rt2x00/rt2x00dev.c
+++ drivers/net/wireless/ralink/rt2x00/rt2x00dev.c
@@ -88,26 +88,26 @@
 	rt2x00queue_start_queues(rt2x00dev);
 	rt2x00link_start_tuner(rt2x00dev);
 
 	/*
 	 * Start watchdog monitoring.
 	 */
-	rt2x00link_start_watchdog(rt2x00dev);
+	//rt2x00link_start_watchdog(rt2x00dev);
 
 	return 0;
 }
 
 void rt2x00lib_disable_radio(struct rt2x00_dev *rt2x00dev)
 {
 	if (!test_and_clear_bit(DEVICE_STATE_ENABLED_RADIO, &rt2x00dev->flags))
 		return;
 
 	/*
 	 * Stop watchdog monitoring.
 	 */
-	rt2x00link_stop_watchdog(rt2x00dev);
+	//rt2x00link_stop_watchdog(rt2x00dev);
 
 	/*
 	 * Stop all queues
 	 */
 	rt2x00link_stop_tuner(rt2x00dev);
 	rt2x00queue_stop_queues(rt2x00dev);
